utils/calendrical_calculations/src/islamic.rs

```diff
@@ -0,0 +1,271 @@
+    if let Some(x) = saudi_criterion(date) {
+        x
+    } else {
+        false
+    }
+}
+
+// Closest fixed date on or before date when Saudi visibility criterion is held.
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6966>
+pub fn saudi_new_month_on_or_before(date: RataDie) -> RataDie {
+    let last_new_moon =
+        libm::floor((Astronomical::lunar_phase_at_or_before(0.0, date.as_moment())).inner()); // Gets the R.D Date of the prior new moon
+    let age = date.to_f64_date() - last_new_moon;
+    // Explanation of why the value 3.0 is chosen: https://github.com/unicode-org/icu4x/pull/3673/files#r1267460916
+    let tau = if age <= 3.0 && !adjusted_saudi_criterion(date) {
+        // Checks if the criterion is not yet visible on the evening of date
+        last_new_moon - 30.0 // Goes back a month
+    } else {
+        last_new_moon
+    };
+
+    next(RataDie::new(tau as i64), adjusted_saudi_criterion) // Loop that increments the day and checks if the criterion is now visible
+}
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6996>
+pub fn saudi_islamic_from_fixed(date: RataDie) -> (i32, u8, u8) {
+    let crescent = saudi_new_month_on_or_before(date);
+    let elapsed_months =
+        libm::round((crescent - FIXED_ISLAMIC_EPOCH_FRIDAY) as f64 / MEAN_SYNODIC_MONTH) as i64;
+    let year = helpers::i64_to_saturated_i32((div_rem_euclid64(elapsed_months, 12).0) + 1);
+    let month = ((div_rem_euclid64(elapsed_months, 12).1) + 1) as u8;
+    let day = ((date - crescent) + 1) as u8;
+
+    (year, month, day)
+}
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6981>
+pub fn fixed_from_saudi_islamic(year: i32, month: u8, day: u8) -> RataDie {
+    let midmonth = RataDie::new(
+        FIXED_ISLAMIC_EPOCH_FRIDAY.to_i64_date()
+            + libm::floor(((year as f64 - 1.0) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH)
+                as i64,
+    );
+    let first_day_of_month = saudi_new_month_on_or_before(midmonth).to_i64_date();
+
+    RataDie::new(first_day_of_month + day as i64 - 1)
+}
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2076>
+pub fn fixed_from_islamic_civil(year: i32, month: u8, day: u8) -> RataDie {
+    let year = i64::from(year);
+    let month = i64::from(month);
+    let day = i64::from(day);
+
+    RataDie::new(
+        (FIXED_ISLAMIC_EPOCH_FRIDAY.to_i64_date() - 1)
+            + (year - 1) * 354
+            + div_rem_euclid_f64(3.0 + year as f64 * 11.0, 30.0).0 as i64
+            + 29 * (month - 1)
+            + div_rem_euclid_f64(month as f64, 2.0).0 as i64
+            + day,
+    )
+}
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2090>
+pub fn islamic_civil_from_fixed(date: RataDie) -> (i32, u8, u8) {
```

Comment:
more naming inconsistency: this is `islamic_civil...`, but above functions are `...saudi_islamic...`

Reply:
civil and saudi are different

Reply:
the names are from the book